### PR TITLE
[FEATURE] multidomain-domainnames

### LINKED ISSUE #56: [FEATURE] Allow combination of separate domain codes for a multi domain setup
Currently there is no way to use the new `-n <domain code>` [option](https://github.com/ffnord/mesh-announce/pull/55) (thanks for implementing that) in a multi domain setup. Unless you start for every domain a seperate instance of mesh-announce. 

In the example for a multi-domain setup:
`respondd.py -d /opt/mesh-announce/providers -i meshvpn-one -i br-one -i bat-one -b bat-one -i meshvpn-two -i br-two -i bat-two -b bat-two -i meshvpn-three -i br-three -i bat-three -b bat-three`
it seems to be suggested not to run multiple instances of mesh-announce for multi domain. 

I don't know which would be the best way to include this information, but somehow you should be able to do this.

This also extends to other options, but `-n` seems to me like something esential for multi domain.

### PR DESCRIPTION
**Pull request by AiyionPrime**
This implements the use of domaincode-files, which are simple json dictionaries holding batadv_device-names as keys and the related domain as their value.

If no or an empty json-file is provided, the code falls back to known behavior and returns the value of "-n" which was implemented in https://github.com/ffnord/mesh-announce/pull/55.
If used this makes the value of '-n' the fallback value.

This PR fixes #56 and provides half of #53.

The idea of binding bat-devices and domaincodes is shamelessly stolen from fftrier:
https://github.com/freifunktrier/mesh-announce/blob/fftr_mods/providers/nodeinfo/system/domain_code.py

The approach of a json was deliberately chosen in order not to spam the commandline-call of respondd.py with further repeating parameters.

### CODE REVIEW

File: util.py [hunk 1]
```diff
@@ -72,3 +73,14 @@ def ifindex_to_batiface(if_index, batman_ifaces):
...
+def read_domainfile(dcf_path):
+    """Read a json file which holds all currently known assignments of
+       bat interfaces to domains as a dictionary and return it as python dict.
+       Return an empty dict, if the given path was None.
+    """
+    if dcf_path is None:
+        return {}
+    with open(dcf_path, "r") as dc_file:
+        known = json.load(dc_file)
+        return known
```
Comment (mweinelt) [inline util.py:86]:
```suggestion
        retorn json.load(dc_file)

```

File: domainfile.json.example [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	"bat15": "dom15-gw",
+	"bat16": "dom16-gw",
+	"bat17": "dom17-gw",
+	"bat18": "dom18-gw",
+	"bat19": "dom19-gw",
+	"bat20": "dom20-gw",
+	"bat21": "dom21-gw",
+	"bat22": "dom22-gw",
+	"bat23": "dom23-gw",
+	"bat99": "dom99-gw"
```
Comment (mweinelt) [inline domainfile.json.example:17]: What's up with the -gw suffix? Shouldn't that match the domain_code given in the Gluon domain?

Comment (AiyionPrime) [reply domainfile.json.example:17]: That text can, but does not necessarily need to reflect the content of gluons domain_code.
Though for hanover it does, this format reflects, what @margau has currently running, if I'm not wrong.  However we can change this, as it's just the example of how to format the file.

File: util.py (hunk 1 above)
Comment (AiyionPrime) [reply util.py:86]: Thought about logging at that point, but later found that that's not a thing in mesh-announce, yet.

File: domainfile.json.example (hunk 2 above)
Comment (margau) [reply domainfile.json.example]: With FFFFM, we actually have different Names per L2-Domains - some on the node-side, mostly PLZ-based, the technical domain ("dom2"), and another one for the gateway, to make the meshviewer-Domain-Name more meaningful.
Example of our meshviewer config.json:
```json
{
      'domain': 'dom17-gw',
      'name': 'Main-Taunus (Gateway)'
    },
{
      'domain': 'ffffm_65817',
      'name': '65817 Eppstein'
    },
{
      'domain': 'ffffm_65812',
      'name': '65812 Bad Soden am Taunus'
    },

```
Both are in the same L2-Domain, but we have more granular Domain Names (per postal code) with only a single gateway instance.

File: util.py [hunk 3]
```diff
@@ -72,3 +73,14 @@ def ifindex_to_batiface(if_index, batman_ifaces):
...
+
+def read_domainfile(dcf_path):
+    """Read a json file which holds all currently known assignments of
+       bat interfaces to domains as a dictionary and return it as python dict.
+       Return an empty dict, if the given path was None.
+    """
+    if dcf_path is None:
+        return {}
+    with open(dcf_path, "r") as dc_file:
+        known = json.load(dc_file)
```
Comment (mweinelt) [inline util.py:85]: That line is now superfluous.

Comment (AiyionPrime) [reply util.py]: that's correct, thanks.

**Comment by AiyionPrime** [comment]
Ah wait. We lack an entry in the readme file.

**Comment by AiyionPrime** [comment]
I'm fine.

**Comment by TobleMiner** [comment]
Hey, thanks for the patch!

I'm not sure about using a config file for the batman-device <-> domain association though. While the command line arguments have already gotten far too complex I don't feel comfortable pushing just this part to a conffile. I'd prefer to keep it part of the command line arguments for now. Maybe as `-b <batman_iface>[:[<mesh_ipv4>][:<domain code>]]`

**Comment by AiyionPrime** [comment]
The association itself should be watched carefully, as it goes straight against @genofire s efforts of possibly getting rid of batman and use babel instead, irrc.
I'd rather communicate this configuration as a temporary necessity, until good things happen, like @mweinelt s rfc [efforts to find a gluon wide solution](https://github.com/freifunk-gluon/gluon/issues/1974) for this, but realistically this will take quite some time.

I do not really agree with the concept of sticking to the pile of commandlinearguments; however if accepting the PR depends on it, I'll implement it for the domain-code.

If someone could provide an example of a working commandline call with two? bat-devices and their mesh_ipv4s in the syntax you provided, it would help me a lot.

> `-b <batman_iface>[:[<mesh_ipv4>]...]`

**Comment by AiyionPrime** [comment]
> I don't feel comfortable pushing just this part to a conffile. I'd prefer to keep it part of the command line arguments for now. Maybe as `-b <batman_iface>[:[<mesh_ipv4>][:<domain code>]]`

Hey @TobleMiner forget my last request for the commandlinecall, I think I found it in the code.

However, this does not make it easier.
You implemented it like this:
`iface, *mesh_ipv4 = ifspec.split(':')`

While I could hardcode, that the last element of ifspec should be the domaincode, this would make it mandatory and therefore break existing deployments.

We don't use the mesh_ipv4 flag in hanover, are there communities using it multiple times per iface?
Or is the star only there to make the parsing easier and everything but the first argument is dropped?

Gotta go, brb.

**Comment by AiyionPrime** [comment]
Well then, this implemented `[:<domain code>]`.
`-b interface` and `-n dom13` are (except the defaulting behavior for other domains) equal to `-b interface::dom13`.
In this case the mesh_ipv4_override is not set for this interface.

This finally allows the domaincode-file to overwrite the default domain given by `-n`.
And the `-b ::` flags to overwrite per interface on the fly.

Therefore it's: `-n` < `-c` < `-b`.

As last thing to allow a start for a proper config-file, I'd like to change the domaincode-file format, so that it's not a dict of domain-codes anymore, but a dict with one key called "domaincodes" and the current dict nested within as its value.
This would allow to seamlessly rename the domain_code file as well the -c parameter to configfile later on, without breaking domaincode related stuff. What do you think about this proposition @TobleMiner , @mweinelt ?

**Comment by AiyionPrime** [comment]
@TobleMiner I think, I'm done. Review requested.

**Comment by TobleMiner** [comment]
Hi,

> I do not really agree with the concept of sticking to the pile of commandlinearguments; however if accepting the PR depends on it, I'll implement it for the domain-code.

Don't get me wrong here, I don't like it either. But I do prefer it over introducing a config file format that is not capable of representing all configuration options of mesh-announce in a proper structure. If we were to introduce the association file right now it would create either create a breaking change as soon as a proper config file format is intoduced or we would need to keep support for the file around for some time. I'll take it like that but be prepared for me kicking out the association file and replacing it by a proper config for the daemon.

I'll test the PR and merge it shortly

**Comment by TobleMiner** [comment]
LGTM, thanks for impelmenting!

**Pull request merged**